### PR TITLE
[platform test]Platform test enhancement

### PR DESCRIPTION
**Pull request by keboliu**
### Description of PR
Enhance the current platform py test.

Summary:
Fixes # (issue)

### Type of change

- [] Bug fix
- [] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### How did you do it?
1. Add platform daemon test to verify whether all the daemons inside pmon are running as expected.
2. enhance the show interface status test by cross-checking the sfp presence status
3. Add sysfs check for CPU, SFP, enhance the fan speed sysfs check, this is Mellanox specific
4. Add cross-check show sfp presence and sysfs qspf_status test for Mellanox platform

#### How did you verify/test it?
run platform test on Mellanox platform

#### Any platform-specific information?
the sysfs check enhancement is Mellanox platform-specific, 
cross-check show sfp presence and sysfs qspf_status test for Mellanox platform

#### Supported testbed topology if it's a new test case?
Support all topology.

### CODE REVIEW

File: tests/common/devices.py [hunk 1]
```diff
@@ -193,3 +194,33 @@ def get_crm_resources(self):
...
+        daemon_config_file_path = os.path.join('/usr/share/sonic/device', self.facts["platform"], 'pmon_daemon_control.json')
+        
+        try:
+            output = self.shell('cat %s' % daemon_config_file_path)
+            json_data = json.loads(output["stdout"])
+            logging.debug("original file content is %s" % str(json_data))
+            for key in full_daemon_tup:
+                if (daemon_ctl_key_prefix + key) not in json_data:
+                    daemon_list.append(key)
+                    logging.debug("daemon %s is enabled." % key)
```
Comment (liat-grozovik) [inline tests/common/devices.py]: suggest to remove the . at the end of the debug line

File: tests/common/devices.py [hunk 2]
```diff
@@ -193,3 +194,33 @@ def get_crm_resources(self):
...
+                elif not json_data[daemon_ctl_key_prefix + key]:
+                    daemon_list.append(key)
+                    logging.debug("daemon %s is enabled." % key)
+                else:
+                    logging.debug("daemon %s is disabled." % key)
+        except:
+            # if pmon_daemon_control.json not exist, by default support all the daemons
```
Comment (liat-grozovik) [inline tests/common/devices.py]: please rephrase. not clear what it means by default support

File: tests/common/mellanox_data.py [hunk 3]
```diff
@@ -17,6 +17,15 @@
...
             "number": 2,
             "hot_swappable": True
+        },
+        "cpu_pack": {
+            "number": 1
+        },
+        "cpu_cores": {
+            "number": 2
+        },
+        "ports":{
```
Comment (liat-grozovik) [inline tests/common/mellanox_data.py]: i am confused on the way number of ports is calculated.
SN2100 is a 16 ports switch unless you breakout the cables, so this is configuration based.
i think you should not have it as hard coded value but instead read the information from the configuration to have the right number.

File: tests/common/mellanox_data.py [hunk 4]
```diff
@@ -122,6 +185,15 @@
...
+            "number": 32
         }
     },
     "ACS-MSN3510": {
```
Comment (liat-grozovik) [inline tests/common/mellanox_data.py]: IMO this system is not part of the supported devices. Please recheck and remove unsupported systems,.

File: tests/platform/check_daemon_status.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        for daemon in daemon_list:
+            output = dut.shell('docker exec -it pmon supervisorctl status | grep %s' % daemon, module_ignore_errors=True)
+            if bool(output["stdout_lines"]):
+                expected_line = output["stdout_lines"][0]
+                expected_line_list = expected_line.split()
+                daemon_status[daemon] = (daemon in expected_line_list and 'RUNNING' in expected_line_list)
+                logging.debug("daemon %s status is %s" % (daemon, str(daemon_status[daemon])))
+            else:
+                logging.debug("daemon %s is not exist" % daemon)
+                #daemon_status[daemon] = False
```
Comment (liat-grozovik) [inline tests/platform/check_daemon_status.py]: looks like code not required. should be removed?

File: tests/common/devices.py [hunk 6]
```diff
@@ -193,3 +194,33 @@ def get_crm_resources(self):
...
+        """
+        full_daemon_tup = ('xcvrd', 'ledd', 'psud', 'syseepromd')
+        daemon_ctl_key_prefix = 'skip_'
+        daemon_list = []
+        daemon_config_file_path = os.path.join('/usr/share/sonic/device', self.facts["platform"], 'pmon_daemon_control.json')
+        
+        try:
+            output = self.shell('cat %s' % daemon_config_file_path)
+            json_data = json.loads(output["stdout"])
+            logging.debug("original file content is %s" % str(json_data))
```
Comment (liat-grozovik) [inline tests/common/devices.py]: need to decide on same methodology in logging information, should it start with capital letter or not. 
if yes, let's make all in the same concept.

File: tests/platform/mellanox/check_sysfs.py [hunk 7]
```diff
@@ -34,46 +34,138 @@ def check_sysfs(dut):
...
             fan_max_content = dut.command("cat %s" % fan_max)
             fan_max_speed = int(fan_max_content["stdout"])
             assert fan_max_speed > 10000, "Bad fan maximum speed: %s" % str(fan_max_speed)
         except Exception as e:
-            assert False, "Get content from %s failed, exception: %s" % (fan_max, repr(e))
+            assert "Get content from %s failed, exception: %s" % (fan_max, repr(e))
+
+        fan_speed_set = "/var/run/hw-management/thermal/fan{}_speed_set".format(fan_id)
+        fan_speed_set_content = dut.command("cat %s" % fan_speed_set)
+        assert fan_speed_set_content["stdout"] == "153", "Fan speed should be set to 60%, 153/255"
```
Comment (liat-grozovik) [inline tests/platform/mellanox/check_sysfs.py]: how is 153 calculated? same for all platforms or should be calculated differently?

File: tests/platform/mellanox/check_sysfs.py [hunk 8]
```diff
@@ -34,46 +34,138 @@ def check_sysfs(dut):
...
+        
+        cpu_pack_max_temp_file = "/var/run/hw-management/thermal/cpu_pack_max"
+        cpu_pack_max_temp_file_output = dut.command("cat %s" % cpu_pack_max_temp_file)
+        cpu_pack_max_temp = float(cpu_pack_max_temp_file_output["stdout"])/1000
+        
+        cpu_pack_crit_temp_file = "/var/run/hw-management/thermal/cpu_pack_crit"
+        cpu_pack_crit_temp_file_output = dut.command("cat %s" % cpu_pack_crit_temp_file)
+        cpu_pack_crit_temp = float(cpu_pack_crit_temp_file_output["stdout"])/1000
+        
+        assert cpu_pack_max_temp <= cpu_pack_crit_temp, "Bad cpu pack max temp or crit temp, %s, %s " % (str(cpu_pack_max_temp), str(cpu_pack_crit_temp))
```
Comment (liat-grozovik) [inline tests/platform/mellanox/check_sysfs.py]: replace cpu w/ CPU in log message.
in log message don't use abbreviations. replace crit with critical

File: tests/platform/mellanox/check_sysfs.py [hunk 9]
```diff
@@ -34,46 +34,138 @@ def check_sysfs(dut):
...
+            psu_fan_speed_get = "/var/run/hw-management/thermal/psu{}_fan1_speed_get".format(psu_id)
+            try:
+                psu_fan_speed_get_content = dut.command("cat %s" % psu_fan_speed_get)
+                psu_fan_speed = int(psu_fan_speed_get_content["stdout"])
+                
+            except Exception as e:
+                assert "Get content from %s failed, exception: %s" % (psu_fan_speed_get, repr(e))
+            assert psu_fan_speed > 1000, "Bad fan speed: %s" % str(psu_fan_speed)
+    
+    sfp_count = SWITCH_MODELS[dut_hwsku]["ports"]["number"]
```
Comment (liat-grozovik) [inline tests/platform/mellanox/check_sysfs.py]: use SFP instead sfp in log message

File: tests/platform/mellanox/test_check_sfp_presence.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+        check_presence_output = ans_host.command(check_intf_presence_command.format(intf))
+        assert check_presence_output["rc"] == 0, "Failed to read interface %s transceiver presence" % intf
+        logging.info(str(check_presence_output["stdout_lines"][2]))
+        presence_list = check_presence_output["stdout_lines"][2].split()
+        logging.info(str(presence_list))
+        assert intf in presence_list, "Wrong interface name in the output %s" % str(presence_list) 
+        assert 'Present' in presence_list, "Status is not expected, output %s" % str(presence_list)
+        
+        check_sysfs_output = ans_host.command(check_qsfp_sysfs_command.format(str(sfp_id)))
+        logging.info('out_put of check sysfs %s' % (str(check_sysfs_output)))
```
Comment (liat-grozovik) [inline tests/platform/mellanox/test_check_sfp_presence.py]: out_put? maybe 'The output'

File: tests/common/devices.py (hunk 1 above)
Comment (keboliu) [reply tests/common/devices.py]: fixed

File: tests/common/devices.py (hunk 2 above)
Comment (keboliu) [reply tests/common/devices.py]: rephrased

File: tests/common/mellanox_data.py (hunk 3 above)
Comment (keboliu) [reply tests/common/mellanox_data.py]: correct the typo of ports number on 2100, this mellanox_data is designed to put never changed platform related info, and check this info against the facts on the DUT, so I think hardcode is OK.

File: tests/common/mellanox_data.py (hunk 4 above)
Comment (keboliu) [reply tests/common/mellanox_data.py]: 3510 not supported yet, removed.

File: tests/platform/check_daemon_status.py (hunk 5 above)
Comment (keboliu) [reply tests/platform/check_daemon_status.py]: removed

File: tests/platform/mellanox/check_sysfs.py (hunk 7 above)
Comment (keboliu) [reply tests/platform/mellanox/check_sysfs.py]: 153 is 60% of 255,  when the hw-mgmt thermal control suspended the fan speed will be set running at 60% speed, this is common for all Mellanox platforms.

File: tests/platform/mellanox/check_sysfs.py (hunk 8 above)
Comment (keboliu) [reply tests/platform/mellanox/check_sysfs.py]: fixed

File: tests/platform/mellanox/check_sysfs.py (hunk 9 above)
Comment (keboliu) [reply tests/platform/mellanox/check_sysfs.py]: fixed

File: tests/platform/mellanox/test_check_sfp_presence.py (hunk 10 above)
Comment (keboliu) [reply tests/platform/mellanox/test_check_sfp_presence.py]: fixed

File: tests/common/devices.py (hunk 6 above)
Comment (keboliu) [reply tests/common/devices.py]: agree, I would keep it unchanged currently, will do it when aligned.

Comment (liat-grozovik) [reply tests/common/devices.py]: Please update the file with one methodology. No reason to push new PR with mixed terminology while new added lines are not stick to one.

File: tests/common/mellanox_data.py (hunk 3 above)
Comment (liat-grozovik) [reply tests/common/mellanox_data.py]: Is it part of the test that you verify it against the DUT? What will happen if the SKU is with split ports?

File: tests/platform/mellanox/check_sysfs.py (hunk 7 above)
Comment (liat-grozovik) [reply tests/platform/mellanox/check_sysfs.py]: OK

File: tests/common/mellanox_data.py (hunk 3 above)
Comment (keboliu) [reply tests/common/mellanox_data.py]: No issue with split ports, check SFP symbolink tests and SFP presence test used this new added ported number, they all work well on the testbeds with split ports.

File: tests/common/devices.py (hunk 6 above)
Comment (keboliu) [reply tests/common/devices.py]: went through the files and they are starting with capital letter, now aligned with it in the newly added files.

**Pull request merged**